### PR TITLE
Bridge RewardTokens from L2 to L1

### PR DESCRIPTION
**Pull request by swapnilraj**

### CODE REVIEW

File: contracts/src/ethereum/TokenBridge.sol
Comment (achab) [inline contracts/src/ethereum/TokenBridge.sol:150]: typo: `payload[2]`instead of `payload[1]`. then increment indices of following lines

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo
Comment (achab) [inline contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:69]: does this one-liner works? usually in cairo, one needs to separate reading from returning ie

```
let (res) = REWARD_TOKEN.read()
return res
```

File: contracts/src/ethereum/TokenBridge.sol
Comment (achab) [inline contracts/src/ethereum/TokenBridge.sol:157]: typo: `receiveRewards` instead of `recieveRewards`

File: contracts/src/starknet/rewaave/token_bridge.cairo
Comment (achab) [inline contracts/src/starknet/rewaave/token_bridge.cairo:155]: shouldn't we burn the reward token here?

Comment (swapnilraj) [reply contracts/src/starknet/rewaave/token_bridge.cairo:155]: This change will be overriden (fix) when merged with your changes to token bridge

File: contracts/src/ethereum/TokenBridge.sol
Comment (swapnilraj) [reply contracts/src/ethereum/TokenBridge.sol:150]: This change will be overriden (fixed) when merged with your changes to token bridge

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo
Comment (JorikSchellekens) [reply contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:56]: These are fine, starkware added this feature a little while ago. The return of a function is a tuple-like thing (being vague here because they are too, tuples are not a native type but they're technically the return type of functions.. 🤷) So you can just return a function call with the same return type.

Comment (achab) [reply contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:49]: Thanks, I wasn't aware of that - and indeed I should have returned (token) in the snippet above.

File: contracts/src/starknet/rewaave/token_bridge.cairo
Comment (achab) [inline contracts/src/starknet/rewaave/token_bridge.cairo:151]: `bridge_address` not used in the function

File: contracts/src/ethereum/TokenBridge.sol
Comment (achab) [reply contracts/src/ethereum/TokenBridge.sol:149]: Not sure it will be fixed as this function is totally new, and won't be merged to a previous implementation.

File: contracts/src/ethereum/TokenBridge.sol
Comment (achab) [inline contracts/src/ethereum/TokenBridge.sol:158]: Shouldn't we switch the lines 157 and 158?

File: contracts/src/ethereum/TokenBridge.sol
Comment (swapnilraj) [reply contracts/src/ethereum/TokenBridge.sol]: I think we should keep it as a separate function but I am not super hell bent on it. In the case we merge it with consumeMessage we'll have to pass in the value for payload[0] and do an if for emitting the right event.
Do you think we should merge it?

File: contracts/src/starknet/rewaave/token_bridge.cairo
Comment (swapnilraj) [reply contracts/src/starknet/rewaave/token_bridge.cairo]: ack

File: contracts/src/ethereum/TokenBridge.sol
Comment (swapnilraj) [reply contracts/src/ethereum/TokenBridge.sol:158]: I think we should consume any messages from l2 first incase their somehow was a message with the recipient set to 0.

Comment (achab) [reply contracts/src/ethereum/TokenBridge.sol:158]: Indeed, I agree with this logic

File: contracts/src/starknet/rewaave/token_bridge.cairo
Comment (hsouf) [inline contracts/src/starknet/rewaave/token_bridge.cairo:101]: Called address should be {governor_}

**Comment by hsouf** [comment]
Let's wait for this PR to be merged first before merging #20

**Review by hsouf — APPROVED** [review]

File: contracts/src/ethereum/L1Token.sol
Comment (achab) [inline contracts/src/ethereum/L1Token.sol:8]: `RewAAVE public REWARD_TOKEN`

File: contracts/src/starknet/rewaave/token_bridge.cairo
Comment (swapnilraj) [reply contracts/src/starknet/rewaave/token_bridge.cairo]: ack

File: contracts/src/starknet/rewaave/token_bridge.cairo
Comment (swapnilraj) [reply contracts/src/starknet/rewaave/token_bridge.cairo:32]: Yea you're right but shouldn't the snake case be `rew_aave`?

File: test/ETHStaticAToken.test.ts
Comment (achab) [inline test/ETHStaticAToken.test.ts:10]: This variable isn't used in the file

File: test/tokenbridge.test.ts
Comment (achab) [inline test/tokenbridge.test.ts:50]: We should probably have a consistent naming between L1 and L2. We could rename `rewAaveToken` with `rewAaveTokenL2`, and `rewAaveL1` with `rewAaveTokenL1`.

File: contracts/src/ethereum/L1Token.sol
Comment (achab) [inline contracts/src/ethereum/L1Token.sol:16]: This function is to be implemented in this PR or in another one?

Comment (swapnilraj) [reply contracts/src/ethereum/L1Token.sol:16]: This function only exists to satisfy the signature for the tests.

File: contracts/src/ethereum/L1Token.sol
Comment (achab) [inline contracts/src/ethereum/L1Token.sol:12]: The contract `L1Token` represents any ERC20 that is accepted by Aave on L1 and L2, right? Those contracts won't have a reward token stored as storage variable. We could instead store it on L1 TokenBridge.

File: contracts/src/ethereum/L1Token.sol
Comment (achab) [reply contracts/src/ethereum/L1Token.sol:16]: Ok, understood!

**Comment by achab** [comment]
Thanks a lot for this PR @swapnilraj, and for having enriched the tests! Once the small comments I made are addressed, you can safely merge it.

File: contracts/src/ethereum/L1Token.sol
Comment (swapnilraj) [reply contracts/src/ethereum/L1Token.sol:12]: I thought L1Token here represents the minimal version of StaticATokenLM that we need for testing.
We only need to represent static tokens since the bridge is intended only for static tokens.

Comment (achab) [reply contracts/src/ethereum/L1Token.sol:12]: You're totally right, I got confused. It's indeed stored like that on `StaticATokenLM`.

File: test/tokenbridge.test.ts
Comment (swapnilraj) [reply test/tokenbridge.test.ts]: fixed

File: test/ETHStaticAToken.test.ts
Comment (swapnilraj) [reply test/ETHStaticAToken.test.ts]: removed

File: contracts/src/ethereum/L1Token.sol
Comment (swapnilraj) [reply contracts/src/ethereum/L1Token.sol]: fixed

**Pull request merged**
